File: models/src/v1/modules/sig_reg.rs
use burn::{
    config::Config,
    module::Module,
    tensor::{Int, Tensor, backend::Backend, linalg::l2_norm, s},
};

use crate::v1::{V1_SIG_REG_POINTS, V1_SIG_REG_SLICES, V1_SIG_REG_T_MAX};

#[derive(Config, Debug)]
pub struct SIGRegConfig {
    #[config(default = "V1_SIG_REG_T_MAX")]
    pub t_max: f64,
    #[config(default = "V1_SIG_REG_POINTS")]
    pub points: usize,
    #[config(default = "V1_SIG_REG_SLICES")]
    pub slices: usize,
}

#[derive(Module, Debug)]
pub struct SIGReg<B: Backend> {
    pub t: Tensor<B, 1>,
    pub phi: Tensor<B, 1>,
    pub weights: Tensor<B, 1>,
    pub slices: usize,
}

impl SIGRegConfig {
    pub fn init<B: Backend>(&self, device: &B::Device) -> SIGReg<B> {
        let indices = Tensor::<B, 1, Int>::arange(0..self.points as i64, device); // step defaults to 1
        let t = indices.float() * (self.t_max / ((self.points - 1) as f64));
        let dt = self.t_max / ((self.points - 1) as f64);

        let weights = Tensor::full([self.points], dt, device);
        let weights =
            weights.slice_assign(s![1..-1], Tensor::full([self.points - 2], dt * 2.0, device));
        let phi: Tensor<B, 1> = (-t.clone().square() / 2.0).exp();
        let weights = weights * phi.clone();
        SIGReg { t, phi, weights, slices: self.slices }
    }
}

impl<B: Backend> SIGReg<B> {
    pub fn forward(&self, input: Tensor<B, 3>, noise: Tensor<B, 2>) -> Tensor<B, 1> {
        let a = noise.clone() / l2_norm(noise, 0);
        let x = input.matmul(a.unsqueeze());
        let n = *x.shape().last().unwrap() as f32;
        let x_t: Tensor<B, 4> = x.unsqueeze_dim(3);
        let x_t = x_t * self.t.clone().unsqueeze();

        let cos_mean: Tensor<B, 3> = x_t.clone().cos().mean_dim(2).squeeze_dim(2);
        let sin_mean: Tensor<B, 3> = x_t.sin().mean_dim(2).squeeze_dim(2);

        let err = ((cos_mean - self.phi.clone().unsqueeze()).powf_scalar(2.0))
            + (sin_mean.powf_scalar(2.0));

        let integrated = err * self.weights.clone().unsqueeze();
        let integrated = integrated.sum_dim(2) * n;

        integrated.mean()
    }
}

File: python-models/src/soma_models/torch/v1/sig_reg.py
import torch
import torch.nn as nn
from dataclasses import dataclass
from soma_models.config import V1_SIG_REG_T_MAX, V1_SIG_REG_POINTS, V1_SIG_REG_SLICES


@dataclass
class SIGRegConfig:
    t_max: float = V1_SIG_REG_T_MAX
    points: int = V1_SIG_REG_POINTS
    slices: int = V1_SIG_REG_SLICES


class SIGReg(nn.Module):
    def __init__(self, config: SIGRegConfig):
        super().__init__()
        t = torch.linspace(0, config.t_max, config.points)
        dt = config.t_max / (config.points - 1)
        weights = torch.full((config.points,), 2 * dt)
        weights[0] = dt
        weights[-1] = dt
        phi = torch.exp(-0.5 * (t**2))
        self.register_buffer("t", t)
        self.register_buffer("phi", phi)
        self.register_buffer("weights", weights * phi)

    def forward(self, x: torch.Tensor, noise: torch.Tensor) -> torch.Tensor:
        A = noise / torch.norm(noise, dim=0)
        x = x @ A
        N = x.shape[-1]
        x_t = x.unsqueeze(-1) * self.t
        cos_mean = torch.mean(torch.cos(x_t), dim=-2)
        sin_mean = torch.mean(torch.sin(x_t), dim=-2)
        err = ((cos_mean - self.phi) ** 2) + (sin_mean**2)
        integrated = torch.sum(err * self.weights, dim=-1) * N
        return integrated.mean()
